src/agent/communicator/src/communicator.cpp

```diff
@@ -100,4 +107,18 @@ namespace communicator
         return res;
     }
 
+    const long long Communicator::GetTokenRemainingSecs() const
+    {
+        auto now = std::chrono::system_clock::now();
+        auto now_seconds = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
+
+        return m_tokenExpTimeInSeconds - now_seconds;
+    }
+
+    void Communicator::WaitForTokenExpirationAndAuthenticate(std::condition_variable& condition, std::mutex& mtx)
```

Comment:
I think we need to find a different approach, let's talk about this.

Reply:
Applied changes according to our conversation.